Add infrastructure for grouping packages list by package level type (top-level or transitive) in the PMUI

jebriede opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/nuget/home/issues/10748
Fixes: https://github.com/nuget/home/issues/10749

Regression? Last working version: N/A

## Description
Adding the infrastructure to group the list of packages in the PMUI by their package level type, whether they are top-level package or transitive packages. This code gets the UI ready for https://github.com/nuget/home/issues/10751, https://github.com/nuget/home/issues/10752, and https://github.com/nuget/home/issues/11486.

The changes in this PR are behind an experimentation flag and will not run without the experiment being active. Without further changes, the code will run once the experiment is active and the aforementioned tasks are completed.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

jebriede committed:
Add infrastructure for grouping packages list by package level type (top-level or transitive) in the PMUI

jebriede commented:
In addition to the unit test added for the value converter, I did manual testing by mocking the top-level packages and making half of them be marked as TopLevel and the other half Transitive. This test was done to check the behavior in the UI and the mocking was reverted before submitting this PR. The result looks like this in its default state (with Transitive collapsed and Top-level expanded by default, but either can be toggled), on the Installed tab. Note that all other tabs do not have grouping which is by design. (Also, the left portion of the screenshot is slightly cut off but the actual UI renders as you would expect):
![image](https://user-images.githubusercontent.com/10777837/150625303-b6c30dc1-5b85-4ae0-8b20-41684eefa7c3.png)

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/PackageLevelTypeToGroupNameConverter.cs:
```diff
@@ -0,0 +1,47 @@
+
+        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
+        {
+            if (value is string pkgLevelTypeString)
```
How is `ConvertBack` being used? What happens if you don't implement it?

jebriede replied:
@donnie-msft it's not being used in this PR. It was implemented for completeness and backed by tests. It may be used in a subsequent PR, depending on how sorting is handled, but it may not.

donnie-msft replied:
Ok, I suggest removing it for now. It doesn't look like it would be difficult to add back if it's needed in the future.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -865,7 +866,7 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
         }
 
-        private async Task<bool> GetUseRecommendedPackagesAsync(PackageLoadContext loadContext, string searchText)
+        private bool GetUseRecommendedPackagesAsync(PackageLoadContext loadContext, string searchText)
```
naming: probably doesn't need to be async anymore.

jebriede replied:
Good catch! It should definitely not have the Async suffix anymore.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -1123,6 +1123,10 @@ private void Filter_SelectionChanged(object sender, FilterChangedEventArgs e)
                     await SearchPackagesAndRefreshUpdateCountAsync(useCacheForUpdates: true);
                     EmitRefreshEvent(timeSpan, RefreshOperationSource.FilterSelectionChanged, RefreshOperationStatus.Success, isUIFiltering: false);
                     _detailModel.OnFilterChanged(e.PreviousFilter, _topPanel.Filter);
+                    if (NuGetExperimentationService.IsExperimentEnabled(ExperimentationConstants.TransitiveDependenciesInPMUI))
```
Can we make this check only once?

nkolev92 reviewed (approved):
Looks good. 

Consider this approve with suggestions.

ghost commented:
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 7 days of this comment**. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.

jebriede committed:
Now adding and clearing grouping when refreshing package list from the consumer of the InfiniteScrollList. Also addressed PR feedback

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -135,10 +136,14 @@ private async ValueTask InitializeAsync(PackageManagerModel model, INuGetUILogge
             _packageList.CheckBoxesEnabled = _topPanel.Filter == ItemFilter.UpdatesAvailable;
             _packageList.IsSolution = Model.IsSolution;
 
+            NuGetExperimentationService = await ServiceLocator.GetComponentModelServiceAsync<INuGetExperimentationService>();
```
I need to remove this line.

jebriede committed:
Removing extra setting of experimentation service

jebriede commented:
@donnie-msft and @nkolev92 thank you for your review so far! I believe I have addressed all your comments. Additionally, in doing further testing, I found some strange behavior with the grouping when switching tabs and refreshing the packages list. I've addressed those issues in these latest commits, so a few things changed:

I removed the concept of what filter is active from the InfiniteScrollList since the control doesn't know what an active tab is today, and I didn't want to introduce that notion unless I had to.
The InfiniteScrollList has methods that can be called by a consumer to Add or Clear the grouping. It's up to the caller to decide when to do those things. This gives the caller more control over when grouping is added and cleared when things outside of the InfiniteScrollList change, such as when a particular tab is selected.
This also lets the consumer of the InfiniteScrollList be in charge of calling these methods only if the experiment is active.
I also added a property on the list to continue using virtualization if grouping is on. Without that property, WPF turns off virtualization for the list when it's grouped which affects its scrolling behavior, among other things.

The grouping behavior works well now. When a tab is changed, the group headers immediately go away so there's no group header artifacts when switching to a tab other than the Installed tab, such as the Browse tab, for example, and grouping is added after the packages list is loaded, only if the Installed tab is selected.

The PR should now be ready for your review again. Thanks in advance!

dominoFire commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml:
```diff
@@ -268,6 +272,33 @@
+    <ControlTemplate
+      x:Key="GroupedListItemControlTemplate"
+      TargetType="{x:Type GroupItem}">
+      <Expander Header="{Binding Name, Converter={StaticResource PackageLevelTypeToGroupNameConverter}}" Style="{StaticResource ExpanderStyle}">
```
Would be useful to include the number of installed/transitive packages in Group Header Title ?

jebriede replied:
This is planned for the next iteration.

dominoFire commented on src/NuGet.Clients/NuGet.VisualStudio.Common/Experimentation/ExperimentationConstants.cs:
```diff
@@ -27,5 +27,6 @@ internal ExperimentationConstants(string flightFlag, string flightEnvironmentVar
         public static readonly ExperimentationConstants PackageManagerBackgroundColor = new("nuGetPackageManagerBackgroundColor", "NUGET_PACKAGE_MANAGER_BACKGROUND_COLOR");
         public static readonly ExperimentationConstants PackageRecommender = new("nugetrecommendpkgs", "NUGET_RECOMMEND_PACKAGES");
         public static readonly ExperimentationConstants BulkRestoreCoordination = new("nugetBulkRestoreCoordination", "NUGET_BULK_RESTORE_COORDINATION");
+        public static readonly ExperimentationConstants TransitiveDependenciesInPMUI = new("nugetTransitiveDependenciesInPMUI", "NUGET_TRANSITIVE_DEPENDENCIES_IN_PMUI");
```
Please rebase to include flag from dev branch.

jebriede replied:
Unless you named the constant differently, the automatic merge should take care of it.

dominoFire commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -716,5 +719,21 @@ public void ResetLoadingStatusIndicator()
+            ItemsView.Refresh();
+            if (ItemsView.OfType<PackageItemViewModel>()?
+                    .Where(p => p.PackageLevelType == PackageLevelType.Transitive)
+                    .Any() == true)
```
```suggestion
            if (ItemsView.OfType<PackageItemViewModel>()
                    .Where(p => p.PackageLevelType == PackageLevelType.Transitive)
                    .Any())
```

dominoFire commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -135,6 +136,9 @@ private async ValueTask InitializeAsync(PackageManagerModel model, INuGetUILogge
             ApplySettings(settings, Settings);
             _initialized = true;
 
+            NuGetExperimentationService = await ServiceLocator.GetComponentModelServiceAsync<INuGetExperimentationService>();
```
Why do we need to keep a reference of experimentation service since exp. check happens only once ?

jebriede replied:
Yes, IsRecommenderFlightEnabled depends on it. Could consider refactoring but that would be out of scope of this PR.

dominoFire commented on src/NuGet.Clients/NuGet.VisualStudio.Common/PackageLevelType.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace NuGet.VisualStudio
+{
+    public enum PackageLevelType
```
I would remove `Type` suffix, since `enum` type suggest a finite, categorical variable.

```suggestion
    public enum PackageLevel
```

dominoFire commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -846,6 +855,11 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
 
+                if (_isTransitiveDependenciesExperimentEnabled && ActiveFilter == ItemFilter.Installed)
+                {
+                    _packageList.AddPackageLevelTypeGrouping();
```
I think some of the UI state logic can be moved to XAML.

One way to accomplish that would be to create a Property and then, add a DataTrigger to set Control Template or Style to with desired UI.

See https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageItemDeprecationLabel.xaml

jebriede replied:
We need more control over when the grouping gets cleared and added than that would offer.

dominoFire reviewed (approved):
Signing-off. Some comments on code style and one on UI.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -716,5 +719,21 @@ public void ResetLoadingStatusIndicator()
+
+        internal void AddPackageLevelTypeGrouping()
+        {
+            ItemsView.Refresh();
```
I see this is a heavier refresh, but I think it's fine since it will only occur when first switching to a tab that supports grouping.

jebriede replied:
As noted, this only occurs when needed to check if there are any transitive dependencies in the list before adding grouping.

donnie-msft reviewed (approved):
Looks good! If you could please comment on any accessibility testing you've done. I'd just want to make sure this meets standards. My first thoughts:

- Theming and color contrasts
- Keyboard accessibility of grouping and toggling
- Screen-reader reads grouping headers/items appropriately
- AccessibilityInsights is clean

nkolev92 reviewed (approved):
LGTM. 

Fernando has a some good feedback I agreed with that is worth addressing (I thumbsed up the ones I think are worth fixing).

jebriede committed:
Address PR feedback

jebriede committed:
Add style to expander that is theme-friendly

jebriede commented:
> Looks good! If you could please comment on any accessibility testing you've done. I'd just want to make sure this meets standards. My first thoughts:
> 
> * Theming and color contrasts
> * Keyboard accessibility of grouping and toggling
> * Screen-reader reads grouping headers/items appropriately
> * AccessibilityInsights is clean

Theming and color contrasts: ✅
Keyboard accessibility of grouping and toggling: ✅
Screen-reader reads grouping headers/items appropriately: ✅
AccessibilityInsights is clean: ✅ (For the newly added groups. AI is not clean for the packages list but that is a problem already present and out of scope of this PR to fix)
